File: day6/day6a.js
const fs = require('node:fs');
const input = fs.readFileSync(__dirname + '/input.txt', 'utf8').replace('\r', '').trim();

const matrix = input.split('\n').map((row) => row.split(''));

let direction = [ -1, 0 ];

const width = matrix[0].length;
const height = matrix.length;

const inBound = (x, y) => x >= 0 && y >= 0 && x < width && y < height;

let position = [ 0, 0 ];
for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
        if (matrix[x][y] === '^') {
            matrix[x][y] = 'X';
            position = [ x, y ];
        }
    }
}

let visited = 1;
let found = false;

while(found === false) {
    while(true) {
        const newX = position[0] + direction[0];
        const newY = position[1] + direction[1];

        if (!inBound(newX, newY)) {
            found = true;
            break
        }
    
        if (matrix[newX][newY] === '#') {
            break;
        }

        if (matrix[newX][newY] !== 'X') {
            matrix[newX][newY] = 'X';
            visited++;
        }

        position = [newX, newY];        
    }

    direction = [ direction[1], 0-direction[0] ];
}

console.log(visited);

File: day6_go/day6a.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("input.txt")
	lines := strings.Split(string(input), "\n")

	width := len(lines[0])
	height := len(lines)
	position := []int{0, 0}

	matrix := make([][]rune, len(lines))
	for i := range matrix {
		matrix[i] = make([]rune, len(lines[i]))
		for j := range lines[i] {
			matrix[i][j] = []rune(lines[i])[j]
			if matrix[i][j] == '^' {
				position[0] = i
				position[1] = j
			}
		}
	}

	direction := []int{-1, 0}

	visited := 1
	found := false

	for {
		if found {
			break
		}

		for {
			newPosition := []int{position[0] + direction[0], position[1] + direction[1]}
			if newPosition[0] < 0 || newPosition[1] < 0 || newPosition[0] >= width || newPosition[1] >= height {
				found = true
				break
			}

			if matrix[newPosition[0]][newPosition[1]] == '#' {
				break
			}

			if matrix[newPosition[0]][newPosition[1]] != 'X' {
				matrix[newPosition[0]][newPosition[1]] = 'X'
				visited += 1
			}

			position = []int{newPosition[0], newPosition[1]}
		}
		direction = []int{direction[1], 0 - direction[0]}
	}

	fmt.Println(visited)
}
